File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/volume/preparation/MaxFileVolumeManager.java
package com.lasthopesoftware.bluewater.client.playback.file.volume.preparation;

import com.lasthopesoftware.bluewater.client.playback.file.volume.ManagePlayableFileVolume;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class MaxFileVolumeManager implements ManagePlayableFileVolume {
	private static final Logger logger = LoggerFactory.getLogger(MaxFileVolumeManager.class);

	private final ManagePlayableFileVolume playableFileVolume;
	private float unadjustedVolume = 1;
	private float maxFileVolume = 1;

	public MaxFileVolumeManager(ManagePlayableFileVolume playableFileVolume) {
		this.playableFileVolume = playableFileVolume;
	}

	@Override
	public float setVolume(float volume) {
		unadjustedVolume = volume;
		final float adjustedVolume = maxFileVolume * volume;
		logger.debug("Volume set to " + volume + ", adjusted volume set to " + adjustedVolume);
		playableFileVolume.setVolume(adjustedVolume);
		return adjustedVolume;
	}

	@Override
	public float getVolume() {
		return playableFileVolume.getVolume();
	}

	public void setMaxFileVolume(float maxFileVolume) {
		logger.debug("Max file volume set to " + maxFileVolume);
		this.maxFileVolume = maxFileVolume;
		setVolume(unadjustedVolume);
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/volume/preparation/MaxFileVolumeManager.kt
package com.lasthopesoftware.bluewater.client.playback.file.volume.preparation

import com.lasthopesoftware.bluewater.client.playback.file.volume.ManagePlayableFileVolume
import com.lasthopesoftware.bluewater.shared.lazyLogger
import com.namehillsoftware.handoff.promises.Promise

class MaxFileVolumeManager(private val playableFileVolume: ManagePlayableFileVolume) : ManagePlayableFileVolume {

	companion object {
		private val logger by lazyLogger<ManagePlayableFileVolume>()
	}

	private var unadjustedVolume = 1f
	private var maxFileVolume = 1f
	override fun setVolume(volume: Float): Promise<Float> {
		unadjustedVolume = volume
		val adjustedVolume = maxFileVolume * volume
		logger.debug("Volume set to $volume, adjusted volume set to $adjustedVolume")
		return playableFileVolume.setVolume(adjustedVolume)
	}

	override val volume: Promise<Float>
		get() = playableFileVolume.volume

	fun setMaxFileVolume(maxFileVolume: Float) {
		logger.debug("Max file volume set to $maxFileVolume")
		this.maxFileVolume = maxFileVolume
		setVolume(unadjustedVolume)
	}
}
